⭐  Token creation tool integration with nSPV, UI/UX improvements

lenilsonjr opened the pull request:
Closes #203 
Closes #100
Closes #216
Closes #199
Closes #173
Closes #223

This PR connects our token creation UI to actual RPC calls. It adds various UX/UI and validation improvements to the tool.
It also adds support for token creation outputs to be show as token balances in the dashboard (see worker changes).

That's it! 🎉

piggydoughnut commented on src/store/models/tokenCreation.ts:
```diff
@@ -0,0 +1,37 @@
+import { createModel } from '@rematch/core';
+
```
I dont think we need a separate model for this as token transaction is just a usual TKL transaction, we can just use currentTransaction model. 

We use currentTransaction for token transactions.

piggydoughnut commented on src/electron/worker.js:
```diff
@@ -144,11 +145,10 @@ class BitgoSingleton {
-      //   res[utxo.tokendata.tokenid.reverse().toString('hex')] = utxo.satoshis;
-      // }
     });
     console.groupEnd();
```
can prolly get rid of this one and the one above too. They are not consoling anything :D

piggydoughnut commented on src/components/Dashboard/widgets/TokenDetailWidget.tsx:
```diff
@@ -106,6 +107,13 @@ const MetadataItem = ({ name, value, copyValue }: MetadataItemProps) => (
   const tokenDetail = useSelector(selectCurrentTokenDetail);
 
+  const hasNumberInCollection =
+    tokenDetail.dataAsJson?.arbitraryAsJson?.number_in_collection ||
```
I would put tokenDetail.dataAsJson?.arbitraryAsJson? in a separate variable, just to make the code a bit more concise

piggydoughnut commented on src/components/CreateToken/TokenCreatedTx.tsx:
```diff
@@ -0,0 +1,107 @@
+            An error has ocurred while broadcasting your {tokenTypeName}: {error}.
+          </p>
+          <p>
+            Please confirm that no transaction has been broacast and try again in a few minutes.
```
how will they check? We could add a suggestion, aka check your address txs on explorer and add a link for users - https://explorer.tokel.io/address/RPVRY4T2KkYN6k9AmRCXFJQcUYmpBYB83Q

piggydoughnut commented on src/components/CreateToken/Form.tsx:
```diff
@@ -143,23 +181,20 @@ const CreateTokenForm: React.FC<CreateTokenFormProps> = ({ tokenType }) => {
               readOnly={tokenType === TokenType.NFT}
               placeholder="100,000"
               min={1}
-              max={200000000}
```
I wonder what are the implications of someone creating too many tokens. Will the chain get jammed with txs, since 1 token = 1 tx right?

lenilsonjr replied:
@piggydoughnut yup.

lenilsonjr replied:
Actually it depends. Reading this again 1 token isn’t 1 tx, since one tx can create multiple tokens fungible between themselves.

piggydoughnut commented on src/components/CreateToken/Form.tsx:
```diff
@@ -113,7 +117,41 @@ const CreateTokenForm: React.FC<CreateTokenFormProps> = ({ tokenType }) => {
+      /* eslint no-underscore-dangle: 0 */
+      if (collectionOption.__isNew__) {
+        setFieldValue('id', Math.floor(Math.random() * 999999));
+        setFieldValue('arbitraryAsJson[collection_name]', collectionOption?.label);
```
this doesnt need to be inside the if. Not that its a big deal :)

piggydoughnut commented on package.json:
```diff
@@ -221,6 +222,7 @@
     "eslint-plugin-react": "^7.20.6",
     "eslint-plugin-react-hooks": "^4.0.8",
     "file-loader": "^6.0.0",
+    "husky": "^4.2.5",
```
isnt the latest version like 7? I remember there was some chat about husky, but cant recall details.

lenilsonjr replied:
yup, but our setup is for 4.2.5. I think we should create a separate ticket to upgrade husky + general pre-hook commits

piggydoughnut commented on src/util/theming.ts:
```diff
@@ -15,6 +15,9 @@ const baseTheme = {
     growth: 'hsla(128, 65%, 48%, 1.0)',
     highlight: 'hsla(171, 94%, 61%, 1.0)',
+    modal: {
+      bg: '#212b3b',
```
can we please make it hsla to align with the rest of the color formats?

piggydoughnut reviewed (commented):
looks great, left a few minor comments. 👍🏽

acnebs commented on src/electron/externalMedia.html:
```diff
@@ -0,0 +1,85 @@
+<html>
```
Is this being run through Prettier? It reformatted on my system when ran prettier.

To avoid the same shenanigans we had with `package.json` it would probably be good to ensure `.html` files are being run through the same formatter in the same way across all our systems.

acnebs commented on src/components/_General/buttons.tsx:
```diff
@@ -23,6 +23,18 @@ const getTheme = theme => {
       return `
         background: var(--color-almostBlack);
         border: 1px solid var(--color-lighterBlack);`;
+    case Colors.DANGER:
```
Can we avoid `switch` statements? They make it easy to break stuff and are usually less readable than the alternatives ([ref](https://medium.com/tandemly/whats-wrong-with-the-switch-statement-in-javascript-c560e8ea3c0b), [ref](https://ericleads.wordpress.com/2012/12/11/switch-case-considered-harmful/)).

lenilsonjr replied:
this was here before though. I'll refactor to if/elses

lenilsonjr replied:
joking. I'll turn that into an object.

acnebs replied:
whew you had me there.

lenilsonjr commented on src/util/validators/useTokenCreationSchema.ts:
```diff
@@ -0,0 +1,70 @@
+          .positive()
+          .integer()
+          .max(maxSupply, `not enough ${TICKER} in wallet`),
+        url: yup.string().url('must be a valid URL'),
```
True. But I'm not sure we even support showing IPFS urls like this. Imo we should track this in a different issue so those URLs are also displayed properly in the dApp & explorer.

acnebs replied:
Fair, just wanted to point out (and relevant to our discussion this afternoon) that https://ipfs.io/ipfs/... is just another gateway so no decentralization in our current implementation. 

In fact why are we running a node if those URLs aren't supported?

lenilsonjr replied:
We parse the hash from IPFS rules and use our node to grab the files. We just need to make sure we’re parsing `IPFS://` urls too and we can change this validation.

acnebs replied:
Ok, I've gone ahead and added support for this. 👀

lenilsonjr replied:
awesome 🎉

lenilsonjr commented on src/hooks/useMyCollections.ts:
```diff
@@ -0,0 +1,36 @@
+  const myPubKey = useSelector(selectAccountPubKey);
+
+  const myCollections = useMemo(() => {
+    const myNFTs = Object.values(tokenDetails).filter(
```
I just simplified as much as I could but since we have to make an array of objects unique, it's kinda hard to not iterate twice. suggestions welcome!

lenilsonjr replied:
Ha, clever. I always underestimate how far `reduce` can go. Thanks for throwing some light!

lenilsonjr commented on src/components/CreateToken/ConfirmTokenCreationModal.tsx:
```diff
@@ -43,100 +52,175 @@ const InformationRow = styled(Columns)`
-  const tokenTypeName = useMemo(() => (token.supply === 1 ? 'NFT' : 'token'), [token]);
+  const tokenCreationSchema = useTokenCreationSchema();
+
+  const tokenType = useMemo(() => (token.supply === 1 ? TokenType.NFT : TokenType.TOKEN), [token]);
```
I'm not a huge fan of this from a syntax/separation of concerns perspective, but understand the performance aspect of this choice. We should make a note of it anyhow in Notion.

acnebs replied:
Yep it does start to get a little concentrated but I think if you're using `useMemo` you gotta prioritize perf over maximally clean code, otherwise may as well not use `useMemo` at all.

lenilsonjr commented:
@acnebs addressed your comments.

`Should we not optimistically add the token to your wallet as soon as it is created, without waiting for confirmation?`
I tend to prefer to have the network tell us the state of any transactions. Thoughts @piggydoughnut?

`Not sure I understand why "Collection" fills ID and they're not just the same field, yet if I enter the collection number manually the collection field does not fill with the collection name.`

Collection ID and Collection name serve different purposes. We don't hold state of the full collection in the blockchain just yet, so saving collection name in each NFT is fine (for now). Once we have a better way to handle this (maybe having a token that represents the whole collection), we can drop name and keep just ID. 

I've changed behavior just so typing an existing collection ID sets the correct name.

acnebs commented:
Looks great. Last two things are to change to [this](https://github.com/TokelPlatform/tokel_dapp/pull/213#discussion_r764167592) for performance and rather than individually setting `tokenData` types as identical `Record<>`s, defining that as a type in `token-types.ts` and using it wherever needed.

acnebs reviewed (changes requested):
Looks great. Last two things are to change to [this](https://github.com/TokelPlatform/tokel_dapp/pull/213#discussion_r764167592) for performance and rather than individually setting `tokenData` types as identical `Record<>`s, defining that as a type in `token-types.ts` and using it wherever needed.

acnebs commented:
Ok, fantastic work @lenilsonjr, if nobody objects I will merge this into `feature/token-creation` now. Then we can let the binaries build and try to run them, then merge into `development`.

lenilsonjr commented:
sounds good to me @acnebs 🙌 thanks for the review and committed improvements.
